[docs] Convert migration guide to asciidoc (#82600)
elastic/kibana: docs/developer/plugin/migrating-legacy-plugins-examples.asciidoc

export function plugin(initializerContext: PluginInitializerContext) {
  return new MyPlugin(initializerContext);
}
----

*plugins/my_plugin/(public|server)/plugin.ts*
[source,typescript]
----
import type { Observable } from 'rxjs';
import { first } from 'rxjs/operators';
import { CoreSetup, Logger, Plugin, PluginInitializerContext, PluginName } from 'kibana/server';
import type { MyPluginConfig } from './config';

export class MyPlugin implements Plugin {
  private readonly config$: Observable<MyPluginConfig>;
  private readonly log: Logger;

  constructor(private readonly initializerContext: PluginInitializerContext) {
    this.log = initializerContext.logger.get();
    this.config$ = initializerContext.config.create();
  }

  public async setup(core: CoreSetup, deps: Record<PluginName, unknown>) {
    const isEnabled = await this.config$.pipe(first()).toPromise();
  }
}
----

Additionally, some plugins need to access runtime env configuration

[source,typescript]
----
export class MyPlugin implements Plugin {
  public async setup(core: CoreSetup, deps: Record<PluginName, unknown>) {
    const { mode: { dev }, packageInfo: { version } } = this.initializerContext.env
  }
----

> === Create a {kib} Platform plugin

Reviewer: ```suggestion
=== Creating a {kib} Platform plugin
```

  id: 'id',
  storedFields: ['some', 'fields'],
})
----

after:

[source,typescript]
----
client.asCurrentUser.get({
  id: 'id',
  stored_fields: ['some', 'fields'],
})
----

* the request abortion API changed

All promises returned from the client API calls now have an `abort`
method that can be used to cancel the request.

before:

[source,typescript]
----
const controller = new AbortController();
legacyClient.callAsCurrentUser('ping', {}, {
  signal: controller.signal,
})
// later
controller.abort();
----

after:

[source,typescript]
----
const request = client.asCurrentUser.ping();
// later
request.abort();
----

* it is now possible to override headers when performing specific API
calls.

Note that doing so is strongly discouraged due to potential side effects
with the ES service internal behavior when scoping as the internal or as
the current user.

[source,typescript]
----
const request = client.asCurrentUser.ping({}, { 
  headers: {
    authorization: 'foo',
    custom: 'bar',
  }
});
----

* the new client doesn’t provide exhaustive typings for the response
object yet. You might have to copy response type definitions from the
Legacy Elasticsearch library until

> https://github.com/elastic/elasticsearch-js/pull/970 merged.

Reviewer: ```suggestion
https://github.com/elastic/elasticsearch-js/pull/970 is merged.
```

[source,typescript]
----
// platform provides a few typings for internal purposes
import type { SearchResponse } from 'kibana/server';
type SearchSource = {...};
type SearchBody = SearchResponse<SearchSource>;
const { body } = await client.search<SearchBody>(...);
interface Info {...}
const { body } = await client.info<Info>(...);
----

* Functional tests are subject to migration to the new client as well.

> before:

Reviewer: ```suggestion

Before:
```

[source,typescript]
----
const client = getService('legacyEs');
----

> after:

Reviewer: ```suggestion
After:
```

[source,typescript]
----
const client = getService('es');
----

> Please refer to the

Reviewer: ```suggestion
Refer to the
```

https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/breaking-changes.html[Breaking
changes list] for more information about the changes between the legacy
and new client.

==== Accessing the client from a route handler

Apart from the API format change, accessing the client from within a
route handler did not change. As it was done for the legacy client, a
preconfigured <<scoped-services, scoped client>> bound to an incoming request is accessible using

> `core` context provider:

Reviewer: ```suggestion
the `core` context provider:
```

[source,typescript]
----
router.get(
  {
    path: '/my-route',
  },
  async (context, req, res) => {
    const { client } = context.core.elasticsearch;
    // call as current user
    const res = await client.asCurrentUser.ping();
    // call as internal user
    const res2 = await client.asInternalUser.search(options);
    return res.ok({ body: 'ok' });
  }
);
----

==== Creating a custom client

Note that the `plugins` option is no longer available on the new
client. As the API is now exhaustive, adding custom endpoints using
plugins should no longer be necessary.

The API to create custom clients did not change much:

> before:

Reviewer: ```suggestion
Before:
```

[source,typescript]
----
const customClient = coreStart.elasticsearch.legacy.createClient('my-custom-client', customConfig);
// do something with the client, such as
await customClient.callAsInternalUser('ping');
// custom client are closable
customClient.close();
----

> after:

Reviewer: ```suggestion
After:
```

[source,typescript]
----
const customClient = coreStart.elasticsearch.createClient('my-custom-client', customConfig);
// do something with the client, such as
await customClient.asInternalUser.ping();
// custom client are closable
customClient.close();
----

> If, for any reasons, one still needs to reach an endpoint not listed on

Reviewer: ```suggestion
If, for any reasons, you still need to reach an endpoint not listed on
```

Joi. This package has full TypeScript support out-of-the-box.

[source,typescript]
----
// Legacy config schema
import Joi from 'joi';

new kibana.Plugin({
  config() {
    return Joi.object({
      enabled: Joi.boolean().default(true),
      defaultAppId: Joi.string().default('home'),
      index: Joi.string().default('.kibana'),
      disableWelcomeScreen: Joi.boolean().default(false),
      autocompleteTerminateAfter: Joi.number().integer().min(1).default(100000),
    })
  }
});

// Kibana Platform equivalent
import { schema, TypeOf } from '@kbn/config-schema';

export const config = {
  schema: schema.object({
    enabled: schema.boolean({ defaultValue: true }),
    defaultAppId: schema.string({ defaultValue: true }),
    index: schema.string({ defaultValue: '.kibana' }),
    disableWelcomeScreen: schema.boolean({ defaultValue: false }),
    autocompleteTerminateAfter: schema.duration({ min: 1, defaultValue: 100000 }),
  })
};

// @kbn/config-schema is written in TypeScript, so you can use your schema
// definition to create a type to use in your plugin code.
export type MyPluginConfig = TypeOf<typeof config.schema>;
----

> ==== Using Kibana Platform config in a new plugin

Reviewer 2: ```suggestion
==== Using {kib} config in a new plugin
```

> After setting the config schema for your plugin, you might want to reach
> the configuration in the plugin. It is provided as part of the

Reviewer 2: ```suggestion
After setting the config schema for your plugin, you might want to read
configuration values from your plugin. It is provided as part of the
```
